### ISSUE TITLE
About the gradient on W

### ISSUE THREAD

**Issue by zhonghp** [wy1iu/sphereface#24]
@wy1iu @ydwen 
In your forward code, you set w to be w' (w' = w/||w||). But in the backward code, you just use ∂E/∂w' as the gradient of w. Actually it should be ∂E/∂w' * ∂w'/∂W, is it a bug?

**Comment by wy1iu** [comment]
No, it is not a bug. We intentionally use the current implementation, but still you could try using ∂E/∂w' * ∂w'/∂W. ∂E/∂w' * ∂w'/∂W should also work in practice. In fact, both using the original gradient and using the current gradient can work perfectly, as long as the objective value can decrease (since the forward pass is exactly what we want). Ours is somewhat similar to projected gradient descent, and is optimized for efficiency and stability, but you could also use the original one.

**Comment by zhonghp** [comment]
@wy1iu Thanks for your reply. 
It seems that the main difference between A-Softmax and L-Softmax is using w' or w, and it makes a big gap in performance. Since I used your modified Softmax (setting m to be 1) and found its performance is poorer than the original Softmax.  I am not very sure what actually make the performance gap between A-Softmax and L-Softmax.

**Comment by wy1iu** [comment]
@zhonghp I believe modified softmax (m=1) performs similarly to softmax. It should not be significantly different as long as you do things right. The performance gain from L-softmax to A-softmax is mainly due to the compatibility between the evaluation metric and the training objective. Because we use cosine similarity in testing, A-softmax is more compatible than L-softmax in terms of the evaluation metric. As you know, L-softmax does not necessarilty optimize the angles (although it does encourage the angular margin), while A-softmax does.
